### PR TITLE
fix(diagnostic): improve diagnostics for swc wasm plugins when mismatch

### PR DESCRIPTION
**Pull request by SyMind**
## Summary

improve diagnostics when wasm plugin binary format mismatch, redirect to https://rspack.dev/errors/swc-plugin-version

image

## Checklist

- [x] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -87,17 +87,15 @@ rspack_dojang = { version = "0.1.9" }
 rate-limit = { existing-packages = 70, new-packages = 70 }
 [profile.dev]
 codegen-units = 16
-debug         = 2       # debug build will cause runtime panic if codegen-unints is default
```
Comment (hardfist) [inline Cargo.toml:90]: no need change this file now

**Comment by h-a-n-a** [comment]
What is the context?

**Comment by chenjiahan** [comment]
I think the motivation for this PR is to add the test case back, @SyMind right?

**Comment by chenjiahan** [comment]
We can change the PR title to make it more accurate.

**Comment by SyMind** [comment]
@hardfist let me revert this PR. Initially, I thought `swc` converted panics into `Result`, which would allow us to modify the error handling. However, after reviewing the swc code, I found that no such change has been made.

Therefore, I believe I misunderstood the purpose of this PR. We need @hardfist to take a look at it.

**Comment by hardfist** [comment]
> What is the context? cc @hardfist @chenjiahan
swc support throw error when wasm binary not match now, so it's safe to add this test case back without enabling catch_on_unwind
https://github.com/swc-project/swc/pull/9562/files#diff-4d01e730818bcc35a17d65dd364390283fd29e3fb1dc0e35050a67034eeb474dR111

File: crates/rspack_loader_swc/src/compiler.rs [hunk 2]
```diff
@@ -346,16 +350,37 @@ impl SwcCompiler {
...
+      match result {
+        Ok(v) => Ok(v),
+        Err(err) => {
+        let error_msg = match err.downcast_ref::<String>(){
+          Some(msg) => {
+            msg
+          },
+          None => "unknown error"
+        };
+        // FIXME: with_help has bugs, use with_help when diagnostic print is fixed
```
Comment (hardfist) [inline crates/rspack_loader_swc/src/compiler.rs:372]: ModuleBuildError will double print help msg if diagnostic contains help, which should be fixed

File: crates/rspack_loader_swc/src/compiler.rs [hunk 3]
```diff
@@ -346,16 +350,37 @@ impl SwcCompiler {
...
+        let help_msg =  indoc!{"
+          The version of the SWC Wasm plugin you're using might not be compatible with `builtin:swc-loader`.
+          The `swc_core` version of the current `rspack_core` is {swc_core_version}. 
+          Please check the `swc_core` version of SWC Wasm plugin to make sure these versions are within the compatible range.
+          Versions of `swc_core` are more likely to be located in the `Cargo.toml` file in the root directory of these plugin repositories.
+          Check out this guide as a reference for selecting the versions of SWC Wasm plugins: https://rspack.dev/errors/swc-plugin-version"};
```
Comment (chenjiahan) [inline crates/rspack_loader_swc/src/compiler.rs:378]:
```suggestion
          See this guide as a reference for selecting SWC Wasm plugin versions: https://rspack.dev/errors/swc-plugin-version"};
```

File: crates/rspack_loader_swc/src/compiler.rs [hunk 4]
```diff
@@ -346,16 +350,37 @@ impl SwcCompiler {
...
+            msg
+          },
+          None => "unknown error"
+        };
+        // FIXME: with_help has bugs, use with_help when diagnostic print is fixed
+        let help_msg =  indoc!{"
+          The version of the SWC Wasm plugin you're using might not be compatible with `builtin:swc-loader`.
+          The `swc_core` version of the current `rspack_core` is {swc_core_version}. 
+          Please check the `swc_core` version of SWC Wasm plugin to make sure these versions are within the compatible range.
+          Versions of `swc_core` are more likely to be located in the `Cargo.toml` file in the root directory of these plugin repositories.
```
Comment (chenjiahan) [inline crates/rspack_loader_swc/src/compiler.rs:377]: Maybe we can remove this line. It is hard for normal JS developers to understand, and we have mentioned this in the FAQ doc.

```suggestion
```

**Review by h-a-n-a — APPROVED** [review]

**Pull request merged**